File: PPImmutableQueue/src/Queue/ImmutableQueue.java
package Queue;

import java.util.NoSuchElementException;

import Stack.IStack;
import Stack.ImmutableStack;

/**
 * ImmutableQueue attempts to create an immutable queue per functional programming paradigms
 * When something new is added to the queue, rather than modify the existing queue a new queue will be made.
 * Conversely, when something is removed from the queue, rather than modify the existing queue a new queue
 * will be made.
 *
 * The class attempts to run an efficiency loop per this article
 * <b>https://blogs.msdn.microsoft.com/ericlippert/2007/12/10/immutability-in-c-part-four-an-immutable-queue/</b>
 * that attempts to optimize de-queueing from the queue by keeping a reverse stack that has the correct reverse
 * order of the queue for faster de-queueing
 *
 * @param <T> Generic Raw Type to initialize and be used to work with the Queue.
 */
public class ImmutableQueue<T> implements IQueue<T>
{
    private static class EmptyQueue<T> implements IQueue<T>
    {
        private static final EmptyQueue emptyQueue = new EmptyQueue();
        static EmptyQueue getInstance()
        {
            return emptyQueue;
        }

        @Override
        public IQueue<T> enqueue(T t)
        {
            return new ImmutableQueue<T>(ImmutableStack.getEmptyStack().push(t), ImmutableStack.getEmptyStack());
        }

        @Override
        public IQueue<T> dequeue()
        {
            throw new NoSuchElementException();
        }

        @Override
        public T head()
        {
            throw new NoSuchElementException();
        }

        @Override
        public boolean isEmpty()
        {
            return true;
        }
    }

    private IStack<T> forwards;
    private IStack<T> backwards;

    private ImmutableQueue(IStack<T> f, IStack<T> b)
    {
        forwards = f;
        backwards = b;
    }

    public static IQueue getNewQueue()
    {
        return EmptyQueue.getInstance();
    }

    @Override
    public IQueue<T> enqueue(T t)
    {
        return new ImmutableQueue<T>(forwards, backwards.push(t));
    }

    @Override
    public IQueue<T> dequeue()
    {
        IStack<T> stack = forwards.pop();
        if(!stack.isEmpty())
        {
            return new ImmutableQueue<T>(stack, backwards);
        }
        else if(backwards.isEmpty())
        {
            return ImmutableQueue.getNewQueue();
        }
        else
        {
            return new ImmutableQueue<T>(ImmutableQueue.reverse(backwards), ImmutableStack.getEmptyStack());
        }
    }

    @Override
    public T head()
    {
        return forwards.head();
    }

    @Override
    public boolean isEmpty()
    {
        return false;
    }

    private static IStack reverse(IStack stack)
    {
        IStack reversed = ImmutableStack.getEmptyStack();
        while(!stack.isEmpty())
        {
            reversed = reversed.push(stack.head());
            stack = stack.pop();
        }
        return reversed;
    }
}

File: ImmutableQueue/src/Queue/ImmutableQueue.scala
package Queue

import Stack.Stack
import Stack.EmptyStack
import Queue.EmptyQueue

class ImmutableQueue[T](f: Stack[T], b: Stack[T]) extends Queue[T] {

  class EmptyQueue[T] extends Queue[T] {
    private final var emptyQueue = EmptyQueue.this
    private [this] def getInstance = emptyQueue

    /**
      * @return if the queue is empty
      */
    override def isEmpty: Boolean = true

    /**
      * @param t Raw Type T to be added to queue
      * @return the Queue.Queue with the element T added
      */
    override def enqueue(t: T): Queue[T] = new ImmutableQueue[T](new EmptyStack[T].push(t), new EmptyStack[T])

    /**
      * @return the Queue.Queue with the first element removed
      */
    override def dequeue(): Queue[T] = throw new NoSuchElementException

    /**
      * @return the Element T at the head of the Queue.Queue
      */
    override def head: Option[T] = throw new NoSuchElementException
  }

  /**
    * @return if the queue is empty
    */
  override def isEmpty: Boolean = false

  /**
    * @param t Raw Type T to be added to queue
    * @return the Queue.Queue with the element T added
    */
  override def enqueue(t: T): Queue[T] = new ImmutableQueue[T](f, b.push(t))

  /**
    * @return the Queue.Queue with the first element removed
    */
  override def dequeue(): Queue[T] = {
    var s = f.pop()
    if(!s.isEmpty) {
       new ImmutableQueue[T](s, b)
    } else if (b.isEmpty) {
       new ImmutableQueue[T](new EmptyStack[T], new EmptyStack[T])
    } else {
       new ImmutableQueue[T](reverse(b), new EmptyStack[T])
    }
  }

  /**
    * @return the Element T at the head of the Queue.Queue
    */
  override def head: Option[T] = Option[T](f.head)

  private def reverse(stack: Stack[T]): Stack[T] = {
    var s = stack
    var reversed = new EmptyStack[T]
    while(!s.isEmpty) {
      reversed = reversed.push(stack.head)
      s = s.pop()
    }
    return reversed
  }
}
